CIPs/0003.md

```diff
@@ -0,0 +1,134 @@
+   * `identifier` - Optional. The identifier being claimed on the service being claimed.  
+   * `data` - Optional. Additional data, e.g. a 
+   URL, to locate a linked signed reference to the claim. 
+* A field array `Claim[] claims` is added to the `Account` struct. 
+
+All claims are optional.
+
+### Primary Claims and Public Metadata
+
+For claims with `service` set to `primary`, `identifier` is a domain name. `data` is an optional path prefix, such that `https://<identifier>/<data>/identity.json` is a file in the following format:
+
+```
+{
+  "address": "<address>",
+  "claim": {
+    "service": "primary",
+    "identifier": "..",
+    "ctime": 1231222312,
+    "expiry" : 31536000, 
+  },
+  "signature": "<sig>",
+}
+```
+
+TODO - detail exact format. 
+
+Clients should verify the SSL/TLS certificate against a chain of trust and check it was issued for the domain and `identifier` value in question. 
+
+Public metadata for a principal with a `primary` claim may be obtained by first verifying the claim then retrieving the resource at `<url>/<address>-metadata.json`.
+
+Example structure of this file:
+
+```
+{
+  "address" : "<address>",
+  "unvalidated": {
+    "name" : "My Name",
+    "organization" : "My Org",
+    "url" : ...,
+    "profile_pic" : ...,
+    "favicon" : ...,
+  },
+  "signature": ...,
+}
+```
+
+TODO - detail exact format. 
+
+### Changes to ContractKit
+
+Add functionality to ContractKit to:
+
+* obtain, verify and cache claims and metadata for an address.
+
+* generate signed identity files ready to be uploaded.
+
+### Changes to Validator elections
+
+Remove the `identifier, name, and url` fields from the `Validator` and `ValidatorGroup` structs in `Validators.sol`, and associated functions.
+
+### Later: Private Metadata
+
+Metadata shared only with the Celo Foundation can be provided at `<url>/<address>-foundation.json.enc` by encrypting a JSON file with the public key of the Celo Foundation.
+
+Example structure of this file:
+
+```
+{
+  "address" : "<address>",
+  "unvalidated": {
+    "phone" : "+11231234567",
+    "whatsapp" : "+11231234567",
+    "email" : "security@example.com",
+  },
+  "signature": ...,
+}
+```
+
+TODO confirm encryption method
+
+### Later: claims for other services
+
+For claims with `service` set to `github`, `identifier` is the Github handle, and `data` identifies the raw contents of a gist, such that `https://gist.githubusercontent.com/<identifier>/<data>` provides a file in the given format. Claim verifiers should take care to disallow relative path segments (e.g `/../`) in `data`. 
+
+For claims with `service` set to `twitter`, `data` is the identifer of a tweet (such that it could be retrieved at `https://api.twitter.com/1.1/statuses/show.json?id=<id>`) whose body has the same format as above.
+
+TODO - twitter requires an API key to retrieve tweets in json format. This is a centralization burden on claim verifiers.
```

Comment:
I think in the worst case, we don't even need to use the API? We could just pull the tweet from the HTML page and grep for the signature?